k8s: Debounce service events

joamaki opened the pull request:
When a service has multiple endpoint slices associated with it this will cause Cilium to process each slice separately, causing repeated updating of the service BPF maps and other churn.

To reduce the processing in these cases, debounce the processing of the service events by collecting events into a buffer every 200ms. If a new event arrives for a specific service during this duration it is coalesced with the prior one.

```release-note
Batch processing of Service and EndpointSlices up to 200 milliseconds to merge repeated changes to a single Service. This significantly reduces the amount of processing Cilium performs for Services with many EndpointSlices.
```

joamaki committed:
k8s: Debounce service events

When a service has multiple endpoint slices associated with it
this will cause Cilium to process each slice separately, causing
repeated updating of the service BPF maps and other churn.

To reduce the processing in these cases, debounce the processing
of the service events by collecting events into a buffer every 200ms.
If a new event arrives for a specific service during this duration it
is coalesced with the prior one.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

giorio94 commented on pkg/k8s/service_cache.go:
```diff
@@ -871,11 +871,11 @@ func (s *ServiceCache) MergeClusterServiceDelete(service *serviceStore.ClusterSe
-			SWGDone:   swg.Add(),
+			Action:       DeleteService,
+			ID:           id,
+			OldService:   svc,
+			OldEndpoints: endpoints,
```
A similar change should be performed in `mergeExternalServiceDeleteLocked` as well, L818.

Additional considerations:
* `clustermesh/services_test.go:314` should ideally be updated as well (it doesn't really matter much, as it only asserts that backends are empty).
* `policy/k8s/service.go:29` is also accessing `Service` and `OldService`. My (limited) understanding of the rest of that logic is that it shouldn't cause problems, because it should not modify the behavior, and all accesses are protected by a nil check. Worth double-checking for confirmation though.

joamaki replied:
Ah thanks for catching it. I somehow glossed over this one even though I checked all refs to Service field. 
Hmm, the OldService isn't set in mergeExternalServiceDeleteLocked which seems wrong. I'll set it, but would appreciate if you double-check that this won't cause surprises (shouldn't, essentially same as EnsureService).

Yep I already checked policy/k8s/service.go. There it's really just triggering the recalculation based on whether a CNP matches or not and it's not using the data passed in the event to actually perform the calculation.

giorio94 replied:
> Hmm, the OldService isn't set in mergeExternalServiceDeleteLocked which seems wrong. I'll set it, but would appreciate if you double-check that this won't cause surprises (shouldn't, essentially same as EnsureService).

Yep, it seems that `mergeServiceUpdateLocked` is potentially setting it to nil as well in certain cases. In terms of service handling it shouldn't matter at all, because `addK8sSVCs` actually treats `svc==oldSvc` and `oldSvc==nil` in the same way (the former just implies slightly more work). Differently, the policy piece would lead to a bit more work when `oldSvc==nil`, as it treats it as a new service. That said, I think that your change makes sense (it could make sense to adapt `mergeServiceUpdateLocked` as well to default it to `svc` when nil, just to ensure consistency). 

I'm definitely happy that this logic is going to be replaced in the future, because there are currently too many caveats and small differences that tracking all corner cases is super difficult right now.

> Yep I already checked policy/k8s/service.go. There it's really just triggering the recalculation based on whether a CNP matches or not and it's not using the data passed in the event to actually perform the calculation.

Perfect, that matches my understanding as well.

giorio94 commented on daemon/cmd/daemon_main.go:
```diff
@@ -490,6 +490,14 @@ func InitGlobalFlags(cmd *cobra.Command, vp *viper.Viper) {
 	option.BindEnv(vp, option.K8sServiceCacheSize)
 	flags.MarkHidden(option.K8sServiceCacheSize)
 
+	flags.Int(option.K8sServiceDebounceBufferSize, 128, "Number of distincts services to buffer for event debouncing")
```
I wonder whether this option should be enabled by default or not, considering that it only benefits medium to large scale environments, to avoid introducing the extra delay on smaller scale ones as well.

joamaki replied:
I would argue that it makes sense to enable it by default to make sure that it is on for the environments that really need this. Otherwise we'll be in a situation where the only way to get this enabled for a user is by having the user of Cilium open a github issue or similar complaining about how badly Cilium behaves.

We could drop the default delay to e.g. 200ms as that's probably already enough to reap the benefits.

giorio94 replied:
Sounds good to me

giorio94 reviewed (approved):
A couple of comments inline, lgtm otherwise. Thanks!

giorio94 commented on operator/watchers/k8s_service_sync.go:
```diff
@@ -42,6 +47,8 @@ func k8sServiceHandler(ctx context.Context, cinfo cmtypes.ClusterInfo, shared bo
 			"action", event.Action,
 			"service", event.Service,
 			"endpoints", event.Endpoints,
+			"old-service", event.Service,
```
Nit. Should be event.OldService.

giorio94 reviewed (approved):
Approving to cover the clustermesh changes as well.

marseel reviewed (approved):
Thanks!

joamaki committed:
k8s: Use Old{Service,Endpoints} in delete events

In preparation for the next commit, make the delete ServiceEvent
carry the data in the OldService and OldEndpoints fields. This is
less confusing and makes it easier to coalesce events when both
the update and delete events carry the "old" state in the Old* fields.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

pippolo84 reviewed (approved):
Nice! :100: 

Nit: I think you need to update the commit message and the release note to account for the latest change in the debounce wait time (`500ms` -> `200ms`)

nathanjsweet reviewed (approved):
Nice.
